### DISCUSSION TITLE
iPerf3 internal mechanism

### DISCUSSION THREAD

**Discussion by keshvadi** [esnet/iperf#1362] [Q&A]
Hi,
I just wanted to know that what is the internal mechanism of iPerf to calculate the available bandwidth? Is it simply saturate the tight link with lots of packets or it uses a technique like packet pair?
And which part of the source code actually handle this mechanism?

I appreciate your time, and thanks for maintaining this tool.

**Comment by davidBar-On** [comment]
Hi, bandwidth is just the amount of data received (by the receiver) or sent (by the sender), divided by the time the data was received/sent.  See the `print_xxxx_results()` functions in [iperf_api.c](https://github.com/esnet/iperf/blob/master/src/iperf_api.c) (search for "bandwidth =").

**Reply by keshvadi** [reply]
Hi David, 
Thanks for your quick reply.
As I know, the sent/received data per second is throughput. Available bandwidth is the maximum remaining capacity of a link or path.
For example, if I send data very slowly, you will receive that data very slowly, the throughput is low (not the available bandwidth).
Thanks for the link. I am going to read that [iperf_api.c](https://github.com/esnet/iperf/blob/master/src/iperf_api.c).

**Reply by bmah888** [reply]
@davidBar-On is correct to the extent that iperf3 is basically just measuring the number of bits sent in a bulk data transfer, divided by time. @Keshvadi is right in that the term for this is throughput.

For tests with TCP, iperf3 will by default send data as fast as it can go. The TCP implementation in the OS kernel will try to send as fast as it can without losing data. So in theory the throughput will approximate the available bandwidth on the channel. You have the option to limit iperf3 to some slower throughput if you like.

UDP doesn't have any reliability mechanisms, so the kernel doesn't have a way to determining how fast it should send. In this case, iperf3 will attempt to send at the speed of the `--bitrate` parameter (default is 1Mbps).

**Reply by keshvadi** [reply]
@bmah888 Hi, thank you for clarification. Thanks for you time
